sql/analyzer/resolve_ctes.go

```diff
@@ -268,103 +285,142 @@ func hoistRecursiveCte(ctx *sql.Context, a *Analyzer, n sql.Node, scope *Scope,
 }
 
 func newRecursiveCte(sq *plan.SubqueryAlias) (sql.Node, error) {
-	// either UNION (deduplicate) or UNION ALL (keep duplicates)
-	var deduplicate bool
-	var union *plan.Union
-	switch n := sq.Child.(type) {
-	case *plan.Distinct:
-		deduplicate = true
-		union = n.Child.(*plan.Union)
-	case *plan.Union:
-		union = n
-	}
-	if union == nil {
-		return nil, sql.ErrInvalidRecursiveCteUnion.New(sq)
-	}
+	union, limit, deduplicate := hoistCTEUnion(sq)
 
 	// TODO: can we support other top-level nodes?
 	// Window, Subquery, RecursiveCte, Cte?
 	switch n := union.Left().(type) {
-	case *plan.Project, *plan.GroupBy:
+	case *plan.Project, *plan.GroupBy, *plan.Having, *plan.Limit:
```

Comment:
Certainly seems like Sort, Filter, etc., should be allowed over here...

Kinda seems like almost anything should be allowed over there? Certainly a UNION on the non-recursive fields works in MySQL:

```sql
WITH RECURSIVE n(i) AS (
   SELECT 1 FROM DUAL
   UNION ALL
   SELECT 2 FROM DUAL
   UNION ALL
   SELECT i + 1 FROM n WHERE i < 10)
SELECT * FROM n
```

And another recursive CTE works:

```sql
WITH RECURSIVE n(i) AS (
   (WITH RECURSIVE f(i) AS (SELECT 1 FROM DUAL UNION SELECT i + 1 FROM f WHERE i < 10) SELECT * FROM f)
   UNION ALL
   SELECT i + 1 FROM n WHERE i < 10)
SELECT * FROM n
```

I realize we just inherited this clause though, so maybe we're not fixing it now.

The behavior we seem to want is:

1) Assert a chain of one or more UNIONs that do not reference the recursive binding.
2) That chain becomes our initial seed and it can do anything.
3) Assert a chain of one or more UNIONs that do reference the recursive binding.
4) That chain becomes our recursive binding and will be run with prior results, accumulating results, until we reach fixed point.

The MySQL I have installed doesn't support 'nesting of unions at the right-hand side', so all unions, including unions in a CTE, must be left binding.

```
if _, ok := union.Right().(*plan.Union); ok {
   return errors.New("This version of MySQL doesn't yet support 'nesting of unions at the right-hand side'")
}
```

In some contexts a user can get around this with a subquery alias

```
select 1 union select * from (select 2 union select 3) a union select 4;
```

but notably that nicely breaks the chain of analysis for this recursive CTE thing and so that one node is either going left or right.

After investigating the the behavior of MySQL recursive CTEs a little, I don't think this stuff is necessarily super far off, but I would kind of expect:

1) `hoistCTEUnion` maybe doesn't return a Union node, which is kind of semantically meaningless and distracting here. Instead, it should return init and recursive themselves. Init should be the Left() sql.Node of the highest Union in the tree that doesn't reference the recursive table. Recursive should be union.Right() of that node in the base case and

```
l, r, limit, distinct := splitRecursiveCTEUnion(n.Left());
NewUnion(r, n.Right())
```

in the recursive case.

It seems like given our parser, at the time we're running `hositCTEUnion`/`splitRecursiveCTEUnion`, we should only have `Limit?(Distinct?(Union(Distinct(Union(...))))` here. The version of MySQL I have installed errors for SORT over UNION in a recursive CTE. It seems like Having and Filter really shouldn't have made their way above the top level Union...?

2) `resolveRecursiveCte` is seemingly restricted in its ability to handle top level nodes partly because of how its getting the output projections...after `default-rules` run on Init, is `Schema()` still liable to panic? Can we punt until it's full resolved, assuming we might need more iterations in some places? Or is there stuff from `Schema()` that we need that isn't in `Schema()`? It seems like it be more natural and much nicer if we have `Init` tell us its fields, map those to the recursive expression fields, replace the recursive table expression in the recursive side and then assert that everything matches up as expected.

Limits and distincts at the top of recursive CTE seem to behave in the way modeled here. In sqlparser.Parse there's a note about how splitting UNION into Distinct(UnionAll(...)). Might make the analyzer's life harder. If I was in this code, I would strongly consider undoing that split, and modeling Distinct(Union(...)) as a thing that can go around as a semantic unit, at least at this point in the analyzer. A once-after rule might still convert any straggling Union(ALL=NO) nodes to Distinct(Union(...)) if they were still around...

Sorry for the brain dump...maybe not at all helpful, but this definitely piqued my interest.

Reply:
thanks! some good ideas about how to tighten this. The subqueries worry me a bit, i designed this for a single scope due to time constraints and it's not immediately clear how we'd sort a subquery into the initial or recursive chain. I suppose we could default it to the recursive part, or search it explicitly for the recursive table name. I can fold all of the other suggestions in tomorrow. My biggest concern is probably still filters.